### PR TITLE
Add comments

### LINKED ISSUE #41: Add comments
Comment should has:
- author
- article
- content

### PR DESCRIPTION
**Pull request by RasimKhusaenov**
### Summary

Added comments.

### How it works

image
image

### Test plan

List of steps to manually test introduced functionality:

- Log in
- Go to https://my-own-blog-pr-62.herokuapp.com/articles/1
- Make sure you can leave a comment and it is displayed

### Deploy notes

Run `bin/rails db:migrate`

### Checklist:

- [x] I have performed a self-review of my own code
- [x] I have added tests for new functionality
- [x] My pull request is shorter than 500 lines

### References
- Resolves #41 Issue

### CODE REVIEW

File: app/views/articles/show.html.slim [hunk 1]
```diff
@@ -1,8 +1,16 @@
...
-p
-  | #{article.content}
+p #{article.content}
 
 - if allowed_to?(:create?, article, with: Users::Articles::PublicationPolicy)
   = button_to "Publish", users_article_publications_path(article), class: "btn btn-primary"
 - if allowed_to?(:destroy?, article, with: Users::Articles::PublicationPolicy)
   = button_to "Back to Draft", users_article_publication_path(article), method: :delete, class: "btn btn-secondary"
+
+- if (current_user.present? && article.published?) || comments.present?
```
Comment (RamilyaNigmatullina) [inline app/views/articles/show.html.slim:9]: What do you think if we always show `Comments` section. In this case we can escape the duplicate conditions.
We can add additional condition to `app/views/comments/_list.html.slim` partial and show "No comments" text if where are no comments. In this case, we can always show comments section and we can always render `comments/_list` partial.

File: db/migrate/20220620135703_create_comments.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class CreateComments < ActiveRecord::Migration[6.1]
+  def change
+    create_table :comments do |t|
+      t.string :content
```
Comment (RamilyaNigmatullina) [inline db/migrate/20220620135703_create_comments.rb:4]: Do we need to add not null constraint?

File: db/migrate/20220620135703_create_comments.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+      t.integer :user_id, null: false
+      t.integer :article_id, null: false
```
Comment (RamilyaNigmatullina) [inline db/migrate/20220620135703_create_comments.rb:6]: Do we need to add index to this field?

**Review by RamilyaNigmatullina — APPROVED** [review]

File: app/controllers/users/comments_controller.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
+module Users
+  class CommentsController < Users::BaseController
+    expose :comment
+    expose :article
+
+    def create
+      self.comment = create_comment.comment
```
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/controllers/users/comments_controller.rb:7]: I plan to gradually switch to the API, so all new features will be implemented through interactors. Or am I wrong?

File: app/policies/users/comment_policy.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+module Users
+  class CommentPolicy < ApplicationPolicy
+    def create?
+      user.present?
```
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/policies/users/comment_policy.rb:4]: There is no article in the new comment, should I attach it to ArticlesController or is there a better solution?

[the remark this answers is not in the document]
Comment (RamilyaNigmatullina) [reply app/policies/users/comment_policy.rb:4]: Comment should have an article, if we use this code
`expose :comment, parent: :article`

File: app/controllers/users/comments_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (RamilyaNigmatullina) [reply app/controllers/users/comments_controller.rb:7]: hm, ok)

File: app/policies/users/comment_policy.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/policies/users/comment_policy.rb:4]: This is not working as expected:
```SQL
Comment Load (0.2ms)  SELECT "comments".* FROM "comments" WHERE "comments"."article_id" = $1 AND "comments"."id" = $2 LIMIT $3  [["article_id", 9], ["id", 9], ["LIMIT", 1]]
```

File: app/controllers/users/comments_controller.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    def authorize_resource!
+      authorize! comment
+    end
+
+    def comment_params
+      params.require(:comment).permit(:content).merge(
+        {
+          user: current_user,
+          article_id: params.require(:article_id)
```
Comment (RamilyaNigmatullina) [inline app/controllers/users/comments_controller.rb:25]: What do you think about this option?
1. use `parent` option of expose
2. pass `comment` to interactor

But you can leave your option if you want.

```ruby
module Users
  class CommentsController < Users::BaseController
    expose :article
    expose :comment, parent: :article

    def create
      if create_comment.success?
        respond_with comment, location: article
      else
        redirect_to article_path(article), alert: create_comment.error
      end
    end

    private

    def authorize_resource!
      authorize! comment
    end

    def comment_params
      params.require(:comment).permit(:content).merge(user: current_user)
    end

    def create_comment
      @create_comment ||= CreateComment.call(comment: comment, comment_params: comment_params)
    end
  end
end
```

File: app/controllers/articles_controller.rb [hunk 7]
```diff
@@ -1,5 +1,7 @@
 class ArticlesController < ApplicationController
   expose :article, scope: -> { authorized_articles }
+  expose :comment, -> { Comment.new(article: article) }
+  expose :comments, from: :article
```
Comment (RamilyaNigmatullina) [inline app/controllers/articles_controller.rb:4]:
```suggestion
  expose :comments, -> { article.comments.includes(:user) }
```

**Review by RamilyaNigmatullina — APPROVED** [review]

File: app/controllers/users/comments_controller.rb (hunk 6 above)
Comment (RasimKhusaenov) [reply app/controllers/users/comments_controller.rb:25]: Oh, that is brilliant! 💎 
Thank you!

File: app/policies/users/comment_policy.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/policies/users/comment_policy.rb]: Never mind. I was doing wrong. :sweat_smile:

File: app/controllers/articles_controller.rb (hunk 7 above)
Comment (RasimKhusaenov) [reply app/controllers/articles_controller.rb:4]: Oh my... Didn't notice it. Thank you!

File: app/controllers/users/comments_controller.rb (hunk 6 above)
Comment (RamilyaNigmatullina) [reply app/controllers/users/comments_controller.rb]: I think, it's ok.

**Review by RamilyaNigmatullina — APPROVED** [review]

File: app/policies/users/comment_policy.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (RamilyaNigmatullina) [reply app/policies/users/comment_policy.rb]: Ok, or we can use this code. `article.comments.new`

[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/policies/users/comment_policy.rb]: Excellent!

**Pull request merged**